File: test_drive/lib/themes/syntax_theme.dart
import 'package:flutter/material.dart';

@immutable
class SyntaxTheme extends ThemeExtension<SyntaxTheme> {
  final Color background;
  final Color defaultText;
  final Color keyword;
  final Color string;
  final Color number;
  final Color comment;
  final Color function;
  final Color typeOfVariable;
  final Color variable;
  final Color operator;
  final Color punctuation;
  final Color attribute;
  final Color tag;

  const SyntaxTheme({
    required this.background,
    required this.defaultText,
    required this.keyword,
    required this.string,
    required this.number,
    required this.comment,
    required this.function,
    required this.typeOfVariable,
    required this.variable,
    required this.operator,
    required this.punctuation,
    required this.attribute,
    required this.tag,
  });

  @override
  SyntaxTheme copyWith({
    Color? background,
    Color? defaultText,
    Color? keyword,
    Color? string,
    Color? number,
    Color? comment,
    Color? function,
    Color? typeOfVariable,
    Color? variable,
    Color? operator,
    Color? punctuation,
    Color? attribute,
    Color? tag,
  }) {
    return SyntaxTheme(
      background: background ?? this.background,
      defaultText: defaultText ?? this.defaultText,
      keyword: keyword ?? this.keyword,
      string: string ?? this.string,
      number: number ?? this.number,
      comment: comment ?? this.comment,
      function: function ?? this.function,
      typeOfVariable: typeOfVariable ?? this.typeOfVariable,
      variable: variable ?? this.variable,
      operator: operator ?? this.operator,
      punctuation: punctuation ?? this.punctuation,
      attribute: attribute ?? this.attribute,
      tag: tag ?? this.tag,
    );
  }

  @override
  SyntaxTheme lerp(
    ThemeExtension<SyntaxTheme>? other,
    double t,
  ) {
    if (other is! SyntaxTheme) return this;

    return SyntaxTheme(
      background: Color.lerp(background, other.background, t)!,
      defaultText: Color.lerp(defaultText, other.defaultText, t)!,
      keyword: Color.lerp(keyword, other.keyword, t)!,
      string: Color.lerp(string, other.string, t)!,
      number: Color.lerp(number, other.number, t)!,
      comment: Color.lerp(comment, other.comment, t)!,
      function: Color.lerp(function, other.function, t)!,
      typeOfVariable: Color.lerp(typeOfVariable, other.typeOfVariable, t)!,
      variable: Color.lerp(variable, other.variable, t)!,
      operator: Color.lerp(operator, other.operator, t)!,
      punctuation: Color.lerp(punctuation, other.punctuation, t)!,
      attribute: Color.lerp(attribute, other.attribute, t)!,
      tag: Color.lerp(tag, other.tag, t)!,
    );
  }

  Color getColorForClass(String? className) {
    if (className == null) return defaultText;

    if (className.contains('keyword')) return keyword;
    if (className.contains('string')) return string;
    if (className.contains('number')) return number;
    if (className.contains('comment')) return comment;
    if (className.contains('function')) return function;
    if (className.contains('type') || className.contains('class')) return typeOfVariable;
    if (className.contains('variable') || className.contains('params')) {
      return variable;
    }
    if (className.contains('operator')) return operator;
    if (className.contains('punctuation')) return punctuation;
    if (className.contains('attr')) return attribute;
    if (className.contains('tag')) return tag;
    if (className.contains('built_in')) return function;
    if (className.contains('literal')) return keyword;
    if (className.contains('symbol')) return variable;
    if (className.contains('title')) return function;

    return defaultText;
  }

  static const SyntaxTheme theme = SyntaxTheme(
    background: Color(0xFF1E1E1E),
    defaultText: Color(0xFFD4D4D4),
    keyword: Color(0xFF569CD6),
    string: Color(0xFFCE9178),
    number: Color(0xFFB5CEA8),
    comment: Color(0xFF6A9955),
    function: Color(0xFFDCDCAA),
    typeOfVariable: Color(0xFF4EC9B0),
    variable: Color(0xFF9CDCFE),
    operator: Color(0xFFD4D4D4),
    punctuation: Color(0xFFD4D4D4),
    attribute: Color(0xFF9CDCFE),
    tag: Color(0xFF569CD6),
  );
}

File: constants/themes/syntaxTheme.ts
import type { SyntaxTheme } from "./types";

function createGetColorForClass(
	theme: Omit<SyntaxTheme, "getColorForClass">,
): SyntaxTheme["getColorForClass"] {
	return (className: string | null) => {
		if (className == null) return theme.defaultText;

		if (className.includes("keyword")) return theme.keyword;
		if (className.includes("string")) return theme.string;
		if (className.includes("number")) return theme.number;
		if (className.includes("comment")) return theme.comment;
		if (className.includes("function")) return theme.function;
		if (className.includes("type") || className.includes("class"))
			return theme.typeOfVariable;
		if (className.includes("variable") || className.includes("params")) {
			return theme.variable;
		}
		if (className.includes("operator")) return theme.operator;
		if (className.includes("punctuation")) return theme.punctuation;
		if (className.includes("attr")) return theme.attribute;
		if (className.includes("tag")) return theme.tag;
		if (className.includes("built_in")) return theme.function;
		if (className.includes("literal")) return theme.keyword;
		if (className.includes("symbol")) return theme.variable;
		if (className.includes("title")) return theme.function;

		return theme.defaultText;
	};
}

// Dark theme (currently only one defined in Flutter)
export const darkSyntaxTheme: SyntaxTheme = {
	background: "#1E1E1E",
	defaultText: "#D4D4D4",
	keyword: "#569CD6",
	string: "#CE9178",
	number: "#B5CEA8",
	comment: "#6A9955",
	function: "#DCDCAA",
	typeOfVariable: "#4EC9B0",
	variable: "#9CDCFE",
	operator: "#D4D4D4",
	punctuation: "#D4D4D4",
	attribute: "#9CDCFE",
	tag: "#569CD6",
	getColorForClass: createGetColorForClass({
		background: "#1E1E1E",
		defaultText: "#D4D4D4",
		keyword: "#569CD6",
		string: "#CE9178",
		number: "#B5CEA8",
		comment: "#6A9955",
		function: "#DCDCAA",
		typeOfVariable: "#4EC9B0",
		variable: "#9CDCFE",
		operator: "#D4D4D4",
		punctuation: "#D4D4D4",
		attribute: "#9CDCFE",
		tag: "#569CD6",
	}),
};

// Light theme (to be defined - using similar colors but adjusted for light background)
export const lightSyntaxTheme: SyntaxTheme = {
	background: "#FFFFFF",
	defaultText: "#24292E",
	keyword: "#005CC5",
	string: "#032F62",
	number: "#005CC5",
	comment: "#6A737D",
	function: "#6F42C1",
	typeOfVariable: "#005CC5",
	variable: "#E36209",
	operator: "#D73A49",
	punctuation: "#24292E",
	attribute: "#005CC5",
	tag: "#22863A",
	getColorForClass: createGetColorForClass({
		background: "#FFFFFF",
		defaultText: "#24292E",
		keyword: "#005CC5",
		string: "#032F62",
		number: "#005CC5",
		comment: "#6A737D",
		function: "#6F42C1",
		typeOfVariable: "#005CC5",
		variable: "#E36209",
		operator: "#D73A49",
		punctuation: "#24292E",
		attribute: "#005CC5",
		tag: "#22863A",
	}),
};
